// Copyright (C) 2018 Bluzelle
//
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU Affero General Public License, version 3,
// as published by the Free Software Foundation.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU Affero General Public License for more details.
//
// You should have received a copy of the GNU Affero General Public License
// along with this program. If not, see <http://www.gnu.org/licenses/>.

syntax = "proto3";

import "database.proto";

message pbft_msg
{
    pbft_msg_type type = 1;

    // used for preprepare, prepare, commit
    uint64 view = 2;
    // used for preprepare, prepare, commit, checkpoint
    uint64 sequence = 3;

    // used for preprepare, prepare, commit, request
    // TODO: Most messages should contain only the hash of the request - KEP-344
    pbft_request request = 4;

    // Needed for prepare, commit, request, checkpoing, but included always for clarity. This is the sender's uuid.
    string sender = 5;

    // for checkpoints
    string state_hash = 6;
}

enum pbft_msg_type {
    PBFT_MSG_UNDEFINED = 0;
    PBFT_MSG_REQUEST = 1;
    PBFT_MSG_PREPREPARE = 2;
    PBFT_MSG_PREPARE = 3;
    PBFT_MSG_COMMIT = 4;
    PBFT_MSG_CHECKPOINT = 5;
}

message pbft_request
{
    database_msg operation = 1;
    uint64 timestamp = 2;
    string client = 3;
}
